src/marine/gmao_obs2ioda.py

```diff
@@ -146,91 +173,71 @@ def __init__(self, files_input, filename, date, obsList):
             varAttrs[(key, metaDataName)]['_FillValue'] = missing_vals[dtypestr]
 
         # Set units and FillValue attributes for groups associated with observed variable.
-        for key in varDict.keys():
-            variable = varDict[key][1]
-            units = varDict[key][2]
-            varAttrs[(variable, obsValName)]['units'] = units
-            varAttrs[(variable, obsErrName)]['units'] = units
-            varAttrs[(variable, obsValName)]['_FillValue'] = float_missing_value
-            varAttrs[(variable, obsErrName)]['_FillValue'] = float_missing_value
+        for key, (_, variable, units, _, _) in presentVarDict.items():
+            for attr_name in [obsValName, obsErrName]:
+                varAttrs[(variable, attr_name)]['units'] = units
+                varAttrs[(variable, attr_name)]['_FillValue'] = float_missing_value
             varAttrs[(variable, qcName)]['_FillValue'] = int_missing_value
 
-        # Skip out if there are no obs!
-        totalObs = 0
-        for obs in obsList:
-            if obs.data['nobs'] <= 0:
-                continue
-            totalObs += obs.data['nobs']
+        # Calculate total number of observations
+        totalObs = sum(obs.data['nobs'] for obs in obsList)
         if totalObs == 0:
             print('No %s observations for IODA!' % varName)
             return
 
-        # data is the dictionary containing IODA friendly data structure
+        # Create data dictionary with IODA friendly data structure
         data = DefaultOrderedDict(lambda: DefaultOrderedDict(dict))
 
-        nlocs = 0
+        for key in meta_keys:
+            dtypestr = locationKeyList[meta_keys.index(key)][1]
+            fill_value = missing_vals[dtypestr]
+            data[(key, metaDataName)] = np.full(totalObs, fill_value, dtype=dtypes[dtypestr])
+
+        print(f"Present variable(s) and their attribute(s) in this input: {presentVarDict}")
+
+        # Prefill data with missing values
+        # QC (preQC) value is zero for now (assume all data is good!)
+        for key, (_, variable, _, _, _) in presentVarDict.items():
+            data[(variable, obsValName)] = np.full(totalObs, float_missing_value, dtype=dtypes['float'])
+            data[(variable, obsErrName)] = np.full(totalObs, float_missing_value, dtype=dtypes['float'])
+            data[(variable, qcName)] = np.full(totalObs, 0, dtype=dtypes['integer'])
```

Comment:
This seems wrong to me, but I may be misunderstanding what is going on. It looks like on line 180 the _FillValue for PreQC is set to `int_missing_value` which I think is good. Looks like on line 64 `int_missing_value` is set to the default integer value which should be the maximum negative 32-bit integer value:  https://github.com/JCSDA-internal/ioda-converters/blob/3c437f464572ce8d5ab076b4ad713f7f7f24dd04/src/pyiodaconv/ioda_conv_engines.py#L51-L52

Then here you set all values in PreQC to zero. Since zero does not match the _FillValue (-2147483647) the ioda reader will not replace the zeros with the JEDI missing values and the UFO operators will be mislead into thinking the data provider for the obs data has marked all values with a zero quality value. Note that in lots of cases a zero quality value means the data is good. 

I think the missing values (`int_missing_value`) should remain in PreQC since PreQC is not given anything extracted from the data provider (ie, the input file to this converter).

My recommendation is to either remove this line (which will leave the missing values, -2147483647, in PreQC), or to replace `0` with `int_missing_value` in this line.

Reply:
So it turns out output from our ODAS processing does not contain QC flags. They are all assumed "good", hence the 0 flag. There is a basic boundary check in this converter script and I added the UFO QC flag "12" if that happens to be triggered. 

In that sense I feel like the code is working as intended, first creating the `_FillValue` attribute, and then assigning `0` for the full `nobs` dimension. But are you saying if I switch from 0 to `int_missing_value` would all of the obs still be used?

Reply:
I agree that the code is doing what you are intending, but I am questioning the use of the PreQC group for this purpose.

The issue I'm bringing up is based on my understanding that any variable in the PreQC group represents QC marks that came from the provider of the data (ie, whoever it was that provided the input file to this converter). If my understanding is correct, then to me this code is speaking on behalf of the provider and inserting good/bad QC without the provider's input. Perhaps another group besides PreQC could be used so it is clear that we have deemed the good/bad quality of the variables and it is clear that the provider hasn't done that.

@BenjaminRuston and @gthompsnJCSDA could you help with this? Please check my understanding and provide guidance. If I'm wrong that's fine but I just want to make sure that we are doing the right thing. Thanks!

Reply:
Based on past usage of quality markers at NCEP-EMC (table 7), I would suggest simply changing the default starting value of PreQC to "2" instead of "0".  That link to quality markers at NCEP indicates that 0 value means it is judged as "keep" and always assimilate, versus 1 for "good" and 2 is used for default or not really checked.

What that means to me is zero is something that may need no additional checks at all (except perhaps `Background Check`) where 2 is saying that UFO should do whatever QC steps are needed before saying the obs is ready for assimilation.

Ultimately, if you believe the incoming data are just always good, then you can pick either zero or one, but I still suggest two is the proper assignment in the converter.